tools/restatectl/src/connection.rs

```diff
@@ -0,0 +1,351 @@
+        kind: MetadataKind,
+        mut guard: MutexGuard<'_, Option<T>>,
+        version_map: M,
+    ) -> Result<T, ConnectionInfoError>
+    where
+        T: StorageDecode + Versioned + Clone,
+        M: Fn(&IdentResponse) -> Version,
+    {
+        if let Some(meta) = &*guard {
+            return Ok(meta.clone());
+        }
+
+        let mut latest_meta: Option<T> = None;
+        let mut answer = false;
+        let mut errors = NodesErrors::default();
+        let mut cache = self.cache.lock().await;
+
+        let request = GetMetadataRequest {
+            kind: kind.into(),
+            sync: false,
+        };
+
+        for address in addresses {
+            let channel = cache.entry(address.clone()).or_insert_with(|| {
+                debug!("connecting to {address}");
+                grpc_channel(address.clone())
+            });
+
+            let mut client = NodeCtlSvcClient::new(channel.clone())
+                .accept_compressed(CompressionEncoding::Gzip)
+                .send_compressed(CompressionEncoding::Gzip);
+
+            let response = match client.get_ident(()).await {
+                Ok(response) => response.into_inner(),
+                Err(status) => {
+                    errors.error(address.clone(), status);
+                    continue;
+                }
+            };
+
+            // node is reachable and has answered
+            answer = true;
+            if response.status != NodeStatus::Alive as i32 {
+                // node did not join the cluster yet.
+                continue;
+            }
+
+            if version_map(&response)
+                <= latest_meta
+                    .as_ref()
+                    .map(|c| c.version())
+                    .unwrap_or(Version::INVALID)
+            {
+                // has older version than we have.
+                continue;
+            }
+
+            let mut response = match client.get_metadata(request).await {
+                Ok(response) => response.into_inner(),
+                Err(status) => {
+                    errors.error(address.clone(), status);
+                    continue;
+                }
+            };
+
+            let meta = StorageCodec::decode::<T, _>(&mut response.encoded)
+                .map_err(|err| ConnectionInfoError::DecoderError(address.clone(), err))?;
+
+            if meta.version()
+                > latest_meta
+                    .as_ref()
+                    .map(|c| c.version())
+                    .unwrap_or(Version::INVALID)
+            {
+                latest_meta = Some(meta);
+            }
+
+            try_best_of -= 1;
+            if try_best_of == 0 {
+                break;
+            }
+        }
+
+        if !answer {
+            // all nodes have returned error
+            return Err(ConnectionInfoError::NodesErrors(errors));
+        }
+
+        *guard = latest_meta.clone();
+        latest_meta.ok_or(ConnectionInfoError::MissingMetadata)
```

Comment:
As a follow-up, it would be great if `restatectl status` or `restatectl nodes ls` or `restatectl meta status` wouldn't fail but maybe list the nodes that are in provisioning state (based on the provided addresses). Here is the corresponding issue: https://github.com/restatedev/restate/issues/2609.